### PR TITLE
Implement predefined acl

### LINKED ISSUE #1660: support for predefined ACL to be set directly on the upload request
This is a follow-up to #651.

This request may be a bit too early, as it will only really matter when #1185 is fixed. Right now with `blob.acl.save_predefined` the object ACL can be set to a predefined value separately, like the _metadata_ can be with `blob.patch`. The idea is to set everything with the upload request, to not have separate request.

### PR DESCRIPTION
**Pull request by chemelnucfin**
Closes #1660

### CODE REVIEW

File: storage/google/cloud/storage/blob.py [hunk 1]
```diff
@@ -688,6 +691,9 @@ def _do_multipart_upload(self, client, stream, content_type,
 
         if self.user_project is not None:
             name_value_pairs.append(('userProject', self.user_project))
+        if predefined_acl is not None:
+            self._acl.save_predefined(predefined_acl)
+            name_value_pairs.append(('predefinedAcl', predefined_acl))
```
Comment (tseaver) [inline storage/google/cloud/storage/blob.py]: I don't think we can use `save_predefined` here -- it forces an API request, which will fail for not-yet-existing blobs.  We likely need to factor out the "name conversion" logic from that method for reuse here.

A system test which exercises the feature would have made this clear.

Comment (tseaver) [reply storage/google/cloud/storage/blob.py]: Also, this argument shouldn't need to be passed down from `upload_*` to the helper methods:  instead, the top-level methods should do the work of setting the ACL:  the lower-level methods will generate the right payload from that.

**Comment by chemelnucfin** [comment]
@tseaver PTAL, thanks.  I set the acl after the blob was created as you had suggested.

**Comment by tseaver** [comment]
@chemelnucfin Moving the `save_predefined` call after the actual upload still makes two API calls.  The change I asked for was to avoid using `save_predefined` at all, and instead to get the appropriate query parameter added to the actual upload API request.  This change will require factoring out the XML->JSON parameter name code (currently inside `save_predefined`), and tweaking `_initiate_resumable_upload` and `_do_multipart_upload` to inject the query parameter appropriately (like for `userProject`).

**Comment by chemelnucfin** [comment]
Sorry, I must have misunderstood you.  I will take a look in a bit.  Thanks.

**Comment by chemelnucfin** [comment]
@tseaver PTAL at your convenience.  Thanks.

File: storage/tests/unit/test_blob.py [hunk 2]
```diff
@@ -1009,9 +1013,11 @@ def _initiate_resumable_helper(
         data = b'hello hallo halo hi-low'
         stream = io.BytesIO(data)
         content_type = u'text/plain'
+        predefined_acl = 'private'
```
Comment (tseaver) [inline storage/tests/unit/test_blob.py]: Shouldn't `_initiate_resumable_helper` take `predefined_acl` as an argument, like `_do_multipart_helper`.  We need to continue to exercise the codepath *without* the value passed in, too.

File: storage/tests/unit/test_blob.py [hunk 3]
```diff
@@ -1180,8 +1187,9 @@ def _do_resumable_helper(self, use_size=False, num_retries=None):
         client = mock.Mock(_http=transport, spec=['_http'])
         stream = io.BytesIO(data)
         content_type = u'text/html'
+        predefined_acl = 'private'
```
Comment (tseaver) [inline storage/tests/unit/test_blob.py]: Likewise `_do_resumable_helper`.

File: storage/tests/unit/test_blob.py [hunk 4]
```diff
@@ -1225,20 +1233,21 @@ def _do_upload_helper(self, chunk_size=None, num_retries=None):
         stream = mock.sentinel.stream
         content_type = u'video/mp4'
         size = 12345654321
-
+        predefined_acl = 'private'
```
Comment (tseaver) [inline storage/tests/unit/test_blob.py]: Likewise `_do_upload_helper`

File: storage/tests/unit/test_blob.py [hunk 5]
```diff
@@ -1252,9 +1261,10 @@ def test__do_upload_with_retry(self):
 
     def _upload_from_file_helper(self, side_effect=None, **kwargs):
         from google.cloud._helpers import UTC
+        from google.cloud.storage.acl import ACL
 
         blob = self._make_one('blob-name', bucket=None)
-
+        blob._acl = mock.create_autospec(ACL)
```
Comment (tseaver) [inline storage/tests/unit/test_blob.py]: Mocking the ACL without making assertions about how it was used seems wrong.

File: storage/tests/unit/test_blob.py [hunk 6]
```diff
@@ -1187,7 +1190,6 @@ def _do_resumable_helper(self, use_size=False, num_retries=None):
         client = mock.Mock(_http=transport, spec=['_http'])
         stream = io.BytesIO(data)
         content_type = u'text/html'
-        predefined_acl = 'private'
         response = blob._do_resumable_upload(
             client, stream, content_type, size, num_retries, None)
```
Comment (tseaver) [inline storage/tests/unit/test_blob.py]: This should be passing through the `predefined_acl` argument, and checking that it is passed through in the assertions.

**Comment by chemelnucfin** [comment]
@tseaver Hopefully this time is good.  PTAL at your convenience.

File: storage/google/cloud/storage/blob.py [hunk 7]
```diff
@@ -941,14 +965,18 @@ def upload_from_file(self, file_obj, rewind=False, size=None,
             warnings.warn(_NUM_RETRIES_MESSAGE, DeprecationWarning)
 
         _maybe_rewind(file_obj, rewind=rewind)
+        predefined_acl = ACL.validate_predefined(predefined_acl)
```
Comment (tseaver) [inline storage/google/cloud/storage/blob.py]: Hmm, I guess that does the Right Thing(TM) when the value passed in is `None`?

File: storage/google/cloud/storage/blob.py [hunk 8]
```diff
@@ -1109,7 +1145,7 @@ def create_resumable_upload_session(
             # to the `ResumableUpload` constructor. The chunk size only
             # matters when **sending** bytes to an upload.
             upload, _ = self._initiate_resumable_upload(
-                client, dummy_stream, content_type, size, None,
+                client, dummy_stream, content_type, size, None, None,
```
Comment (tseaver) [inline storage/google/cloud/storage/blob.py]: For clarity, can we use keyword arguments for the two `None` values here?

Comment (chemelnucfin) [reply storage/google/cloud/storage/blob.py]: They are actually positional args.  Let's just fix this in another PR for clarity purposes.

File: storage/google/cloud/storage/blob.py (hunk 7 above)
Comment (chemelnucfin) [reply storage/google/cloud/storage/blob.py]: Seemed like it and shown with the test.  it was code nicely written which I moved.

**Comment by chemelnucfin** [comment]
@tseaver added the unit test for validation.

**Pull request merged**